apps/prairielearn/src/ee/auth/lti13/lti13Auth.ts

```diff
@@ -0,0 +1,259 @@
+import { Router, type Request, Response, NextFunction } from 'express';
+import asyncHandler = require('express-async-handler');
+import { Issuer, Strategy, type TokenSet } from 'openid-client';
+import * as passport from 'passport';
+import { z } from 'zod';
+import { get as _get } from 'lodash';
+import { callbackify } from 'util';
+
+import { loadSqlEquiv, queryAsync } from '@prairielearn/postgres';
+import * as error from '@prairielearn/error';
+import * as authnLib from '../../../lib/authn';
+import { selectLti13Instance } from '../../models/lti13Instance';
+import { get as cacheGet, set as cacheSet } from '../../../lib/cache';
+import { getInstitutionAuthenticationProviders } from '../../lib/institution';
+
+const sql = loadSqlEquiv(__filename);
+const router = Router({ mergeParams: true });
+
+// Middleware to check access
+router.use(
+  asyncHandler(async (req, res, next) => {
+    const lti13_instance = await selectLti13Instance(req.params.lti13_instance_id);
+    const instAuthProviders = await getInstitutionAuthenticationProviders(
+      lti13_instance.institution_id,
+    );
+
+    if (!instAuthProviders.some((a) => a.name === 'LTI 1.3')) {
+      throw error.make(404, 'Institution does not support LTI 1.3 authentication');
+    }
+
+    next();
+  }),
+);
+
+// Express routes
+router.get('/login', asyncHandler(launchFlow));
+router.post('/login', asyncHandler(launchFlow));
```

Comment:
Is responding to both `POST` and `GET` necessary? I only ask because other auth methods (SAML, Google OAuth, Azure) only need to handle `GET`.

Reply:
https://www.imsglobal.org/spec/security/v1p0/#step-1-third-party-initiated-login

> When a user wants to launch into a Tool, the Platform will start the OpenID Connect flow by redirecting the User Agent (UA) to the 3rd party initiated login end point. The redirect may be a form POST or a GET - a Tool must support either case, with the following parameters.

Reply:
I'm loving these links to the spec! Could you stick this in an inline comment?

Reply:
I have a browser window that's dedicated to the LTI references. URLs I keep blinking back over to are https://www.imsglobal.org/spec/lti/v1p3 and https://www.imsglobal.org/spec/security/v1p0/ as the core pieces. 

There are some extensions relating to names and roles and grading: https://www.imsglobal.org/spec/lti-nrps/v2p0/ and https://www.imsglobal.org/spec/lti-ags/v2p0/